### ISSUE TITLE
libusb example similar to libusbk kbench or libusbdotnet benchmarkcon

### ISSUE THREAD

**Issue by mcuee** [libusb/libusb#962]
It will be nice to have a libusb based example similar to libusbk kbench (or libusbdotnet benchmarkcon). They are not really limited to test for the benchmark FW, but rather generic loopback or sourcesink FW (available for many devices including Linux USB Gadget device).

Ref:
https://github.com/libusb/libusb/issues/951

**Comment by mcuee** [comment]
Example run log under my Mac Mini M1.
```
libusbdotnet_v2 on  v2 took 24s ❯ ./stage/BenchmarkCon/bin/Release/net5.0/BenchmarkCon                                                                      
LibUsbDotNet USB Benchmark v2.2.0.0
Copyright (c) 2010 Travis Robinson. <libusbdotnet@gmail.com>
website: http://sourceforge.net/projects/libusbdotnet

USAGE: benchmark [list] 
                 [pid=] [vid=] [intf=] [altf=] [ep=]
                 [read|write|loop] [notestselect]
                 [verify|verifydetail]
                 [retry=] [timeout=] [refresh=] [priority=]
                 [mode=] [buffersize=] [buffercount=] [packetsize=]
                 [driver=]
                 
Commands:
         list  : Display a list of connected devices before starting. 
                 Select the device to use for the test from the list.
         read  : Read from the device.
         write : Write to the device.
         loop  : [Default] Read and write to the device at the same time.

         notestselect : Skips submitting the control transfers to get/set the
                        test type.  This makes the application compatible
                        with non-benchmark firmwared. Use at your own risk!

         verify       : Verify received data for loop and read tests. Report
                        basic information on data validation errors.
         verifydetail : Same as verify except reports detail information for 
                        each byte that fails validation.
                        
Switches:
         vid        : Vendor id of device. (hex)  (Default=1234)
         pid        : Product id of device. (hex) (Default=5678)
         retry      : Number of times to retry a transfer that timeout.
                      (Default = 0)
         timeout    : Transfer timeout value. (milliseconds) (Default=5000)
                      The timeout value used for read/write operations. If a
                      transfer times out more than {retry} times, the test 
                      fails and the operation is aborted.
         mode       : Sync|Async (Default=Sync) 
                      Sync uses the libusb-win32 sync transfer functions.
                      Async uses the libusb-win32 asynchronous api.
         buffersize : Transfer test size in bytes. (Default=4096)
                      Increasing this value will generally yield higher
                      transfer rates.
         buffercount: (Async mode only) Number of outstanding transfers on
                      an endpoint (Default=1, Max=10). Increasing this value
                      will generally yield higher transfer rates.
         refresh    : The display refresh interval. (in milliseconds)
                      (Default=1000) This also effect the running status.
         priority   : AboveNormal|BelowNormal|Highest|Lowest|Normal
                      (Default=Normal) The thread priority level to use
                      for the test.
         ep         : The loopback endpoint to use. For example ep=0x01, would
                      read from 0x81 and write to 0x01. (default is to use the
                      (first read/write endpoint(s) in the interface)
         intf       : The interface id the read/write endpoints reside in.
         altf       : The alternate interface setting number to select.
         packetsize : For isochronous use only. Sets the iso packet size.
                      If not specified, the endpoints maximum packet size
                      is used.
         driver     : winusb|libusb-win32|libusb10.  Forces the use a specific
                      driver.  If not specified, all drivers are used.
         
WARNING:
          This program should only be used with USB devices which implement
          one more more "Benchmark" interface(s).  Using this application
          with a USB device it was not designed for can result in permanent
          damage to the device.
          
Examples:
benchmark list
benchmark list read altf=1
benchmark vid=1234 pid=5678
benchmark vid=0x4D2 pid=0x162E
benchmark vid=0x4D2 pid=0x162E buffersize=65536
benchmark read vid=0x4D2 pid=0x162E
benchmark vid=0x4D2 pid=0x162E buffercount=3 buffersize=0x2000
```

**Comment by mcuee** [comment]
Results under my Mac Mini M1, for Cypress FX3 BulkSrcSink FW. 

```
libusbdotnet_v2 on  v2 took 18s ❯ ./stage/BenchmarkCon/bin/Release/net5.0/BenchmarkCon list notestselect write buffercount=3 buffersize=32768 loop mode=async
LibUsbDotNet USB Benchmark v2.2.0.0
Copyright (c) 2010 Travis Robinson. <libusbdotnet@gmail.com>
website: http://sourceforge.net/projects/libusbdotnet
1. 04B4:00F1 Cypress - FX3
2. 046D:C52B Logitech - USB Receiver
3. 047F:C025 Plantronics C320-M
4. 1A40:0201 
5. 0AC8:3420 Vimicro Corp. - PROLiNK PCC3220
6. 1A40:0201 
7. 0BDA:5411 Generic - 4-Port USB 2.0 Hub
8. 05E3:0749 Generic - USB3.0 Card Reader
9. 0BDA:0411 Generic - 4-Port USB 3.0 Hub
10. 0BDA:0411 Generic - 4-Port USB 3.0 Hub
11. 0BDA:5411 Generic - 4-Port USB 2.0 Hub

Select device (1-11) :1

Benchmark device 04B4:00F1 opened..
Loop Test Information
	Vid / Pid       : 04B4h / 00F1h
	Interface #     : 00h
	Alt Interface # : 00h
	Priority        : Normal
	Buffer Size     : 32768
	Buffer Count    : 3
	Display Refresh : 1000 (ms)
	Transfer Timeout: 5000 (ms)
	Retry Count     : 0
	Verify Data     : Off

Bulk Read (Ep81h) max packet size: 1024
Bulk Write (Ep01h) max packet size: 1024

While the test is running:
Press 'Q' to quit
Press 'T' for test details
Press 'I' for status information
Press 'R' to reset averages

Press 'Q' to exit, any other key to begin..
Avg. Bytes/s: 281226203.00 Transfers: 8086 Bytes/s: 281226203.00
Avg. Bytes/s: 288942931.94 Transfers: 17199 Bytes/s: 296153452.65
Avg. Bytes/s: 291884573.12 Transfers: 26336 Bytes/s: 297587420.67
Avg. Bytes/s: 293362093.71 Transfers: 35470 Bytes/s: 297707201.40
Avg. Bytes/s: 294178361.38 Transfers: 44600 Bytes/s: 297393131.68
Avg. Bytes/s: 294741105.78 Transfers: 53733 Bytes/s: 297520426.37
Avg. Bytes/s: 295190625.51 Transfers: 62869 Bytes/s: 297862455.43
Avg. Bytes/s: 295488757.12 Transfers: 72000 Bytes/s: 297557922.46
Avg. Bytes/s: 295693625.44 Transfers: 81130 Bytes/s: 297319243.79
Avg. Bytes/s: 295910920.67 Transfers: 90252 Bytes/s: 297857662.74
Avg. Bytes/s: 296073306.84 Transfers: 99385 Bytes/s: 297687636.52
Avg. Bytes/s: 296204760.63 Transfers: 108519 Bytes/s: 297642661.24
Avg. Bytes/s: 296320007.38 Transfers: 117654 Bytes/s: 297695970.59
Avg. Bytes/s: 296410631.73 Transfers: 126781 Bytes/s: 297583831.34
qAvg. Bytes/s: 296493732.55 Transfers: 135920 Bytes/s: 297651376.26
stopped Ep81h thread.
stopped Ep01h thread.
Loop Test Information
	Vid / Pid       : 04B4h / 00F1h
	Interface #     : 00h
	Alt Interface # : 00h
	Priority        : Normal
	Buffer Size     : 32768
	Buffer Count    : 3
	Display Refresh : 1000 (ms)
	Transfer Timeout: 5000 (ms)
	Retry Count     : 0
	Verify Data     : Off

Bulk Read (Ep81h) max packet size: 1024
	Total Bytes     : 4453859328
	Total Transfers : 135921
	Avg. Bytes/sec  : 296490308.47
	Elapsed Time    : 15.02 seconds

Bulk Write (Ep01h) max packet size: 1024
	Total Bytes     : 4822728704
	Total Transfers : 147178
	Avg. Bytes/sec  : 321044531.44
	Elapsed Time    : 15.02 seconds

Press any key to exit..
```

**Comment by mcuee** [comment]
Run log with single direction.

```
libusbdotnet_v2 on  v2 took 5s ❯ ./stage/BenchmarkCon/bin/Release/net5.0/BenchmarkCon list notestselect read buffercount=4 buffersize=32768 loop mode=async
LibUsbDotNet USB Benchmark v2.2.0.0
Copyright (c) 2010 Travis Robinson. <libusbdotnet@gmail.com>
website: http://sourceforge.net/projects/libusbdotnet
1. 04B4:00F1 Cypress - FX3
2. 046D:C52B Logitech - USB Receiver
3. 047F:C025 Plantronics C320-M
4. 1A40:0201 
5. 0AC8:3420 Vimicro Corp. - PROLiNK PCC3220
6. 1A40:0201 
7. 0BDA:5411 Generic - 4-Port USB 2.0 Hub
8. 05E3:0749 Generic - USB3.0 Card Reader
9. 0BDA:0411 Generic - 4-Port USB 3.0 Hub
10. 0BDA:0411 Generic - 4-Port USB 3.0 Hub
11. 0BDA:5411 Generic - 4-Port USB 2.0 Hub

Select device (1-11) :1

Benchmark device 04B4:00F1 opened..
Loop Test Information
	Vid / Pid       : 04B4h / 00F1h
	Interface #     : 00h
	Alt Interface # : 00h
	Priority        : Normal
	Buffer Size     : 32768
	Buffer Count    : 4
	Display Refresh : 1000 (ms)
	Transfer Timeout: 5000 (ms)
	Retry Count     : 0
	Verify Data     : Off

Bulk Read (Ep81h) max packet size: 1024
Bulk Write (Ep01h) max packet size: 1024

While the test is running:
Press 'Q' to quit
Press 'T' for test details
Press 'I' for status information
Press 'R' to reset averages

Press 'Q' to exit, any other key to begin..
Avg. Bytes/s: 273813690.34 Transfers: 7847 Bytes/s: 273813690.34
Avg. Bytes/s: 285690931.32 Transfers: 16955 Bytes/s: 296782138.04
Avg. Bytes/s: 289820435.65 Transfers: 26097 Bytes/s: 297803844.07
Avg. Bytes/s: 291755212.08 Transfers: 35231 Bytes/s: 297428221.92
Avg. Bytes/s: 292893683.41 Transfers: 44357 Bytes/s: 297373395.37
Avg. Bytes/s: 293787759.87 Transfers: 53511 Bytes/s: 298198602.05
Avg. Bytes/s: 294421763.82 Transfers: 62657 Bytes/s: 298186706.89
Avg. Bytes/s: 294871602.07 Transfers: 71802 Bytes/s: 297991034.32
Avg. Bytes/s: 295190648.74 Transfers: 80945 Bytes/s: 297720399.16
Avg. Bytes/s: 295464122.07 Transfers: 90087 Bytes/s: 297907794.94
Avg. Bytes/s: 295719481.83 Transfers: 99238 Bytes/s: 298257134.16
Avg. Bytes/s: 295918548.99 Transfers: 108385 Bytes/s: 298095630.04
Avg. Bytes/s: 296067892.65 Transfers: 117524 Bytes/s: 297850614.57
Avg. Bytes/s: 296207542.31 Transfers: 126665 Bytes/s: 298014796.66
Avg. Bytes/s: 296318649.53 Transfers: 135804 Bytes/s: 297867205.11
Avg. Bytes/s: 296414841.78 Transfers: 144942 Bytes/s: 297851795.26
Avg. Bytes/s: 296487495.72 Transfers: 154077 Bytes/s: 297645056.08
qAvg. Bytes/s: 296564969.26 Transfers: 163232 Bytes/s: 297874936.45
stopped Ep81h thread.
stopped Ep01h thread.
Loop Test Information
	Vid / Pid       : 04B4h / 00F1h
	Interface #     : 00h
	Alt Interface # : 00h
	Priority        : Normal
	Buffer Size     : 32768
	Buffer Count    : 4
	Display Refresh : 1000 (ms)
	Transfer Timeout: 5000 (ms)
	Retry Count     : 0
	Verify Data     : Off

Bulk Read (Ep81h) max packet size: 1024
	Total Bytes     : 5348818944
	Total Transfers : 163233
	Avg. Bytes/sec  : 296562313.60
	Elapsed Time    : 18.04 seconds

Bulk Write (Ep01h) max packet size: 1024
	Total Bytes     : 5807144960
	Total Transfers : 177220
	Avg. Bytes/sec  : 321974803.43
	Elapsed Time    : 18.04 seconds

Press any key to exit..
```

**Comment by mcuee** [comment]
Isoc transfer is not working though, this is probably a libusbdotnet issue.

libusbdotnet_v2 on  v2 took 5s ❯ ./stage/BenchmarkCon/bin/Release/net5.0/BenchmarkCon list notestselect loop buffercount=3 buffersize=16384 mode=async altf=1
LibUsbDotNet USB Benchmark v2.2.0.0
Copyright (c) 2010 Travis Robinson. <libusbdotnet@gmail.com>
website: http://sourceforge.net/projects/libusbdotnet
1. 04B4:00F0 Cypress - FX3
2. 047F:C025 Plantronics C320-M
3. 046D:C52B Logitech - USB Receiver
4. 1A40:0201 
5. 0AC8:3420 Vimicro Corp. - PROLiNK PCC3220
6. 1A40:0201 
7. 05E3:0749 Generic - USB3.0 Card Reader
8. 0BDA:5411 Generic - 4-Port USB 2.0 Hub
9. 0BDA:0411 Generic - 4-Port USB 3.0 Hub
10. 0BDA:5411 Generic - 4-Port USB 2.0 Hub
11. 0BDA:0411 Generic - 4-Port USB 3.0 Hub

Select device (1-11) :1

Benchmark device 04B4:00F0 opened..
Loop Test Information
	Vid / Pid       : 04B4h / 00F0h
	Interface #     : 00h
	Alt Interface # : 01h
	Priority        : Normal
	Buffer Size     : 16384
	Buffer Count    : 3
	Display Refresh : 1000 (ms)
	Transfer Timeout: 5000 (ms)
	Retry Count     : 0
	Verify Data     : Off

Isochronous Read (Ep83h) max packet size: 1024
Isochronous Write (Ep03h) max packet size: 1024

While the test is running:
Press 'Q' to quit
Press 'T' for test details
Press 'I' for status information
Press 'R' to reset averages

Press 'Q' to exit, any other key to begin..
err:failed reading! 1 of 1 ret=MonoApiError: MonoApiError:SubmitTransfer Ep 0x03 
-2:ErrorInvalidParam:Invalid parameter
err:failed writing! 1 of 1 ret=MonoApiError: MonoApiError:SubmitTransfer Ep 0x03 
-2:ErrorInvalidParam:Invalid parameter
stopped Ep83h thread.
stopped Ep03h thread.
Loop Test Information
	Vid / Pid       : 04B4h / 00F0h
	Interface #     : 00h
	Alt Interface # : 01h
	Priority        : Normal
	Buffer Size     : 16384
	Buffer Count    : 3
	Display Refresh : 1000 (ms)
	Transfer Timeout: 5000 (ms)
	Retry Count     : 0
	Verify Data     : Off

Isochronous Read (Ep83h) max packet size: 1024
Isochronous Write (Ep03h) max packet size: 1024
Press any key to exit..

**Comment by mcuee** [comment]
Pretty decent results with my Raspberry Pi 400, running 64bit Ubuntu Linux.

```
libusbdotnet_v2 on  v2 took 18s ❯ ./stage/BenchmarkCon/bin/Release/net5.0/BenchmarkCon list notestselect loop mode=async buffercount=4 buffersize=65536
LibUsbDotNet USB Benchmark v2.2.0.0
Copyright (c) 2010 Travis Robinson. <libusbdotnet@gmail.com>
website: http://sourceforge.net/projects/libusbdotnet
1. 1D6B:0002 
2. 04B4:00F1 Cypress - FX3
3. 05E3:0749 
4. 0BDA:0411 
5. 0BDA:0411 
6. 1D6B:0003 
7. 04D9:0007 
8. 0AC8:3420 
9. 047F:C025 
10. 1A40:0201 
11. 046D:C52B 
12. 1A40:0201 
13. 0BDA:5411 
14. 0BDA:5411 
15. 2109:3431 
16. 1D6B:0002 

Select device (1-16) :2

Benchmark device 04B4:00F1 opened..
Loop Test Information
	Vid / Pid       : 04B4h / 00F1h
	Interface #     : 00h
	Alt Interface # : 00h
	Priority        : Normal
	Buffer Size     : 65536
	Buffer Count    : 4
	Display Refresh : 1000 (ms)
	Transfer Timeout: 5000 (ms)
	Retry Count     : 0
	Verify Data     : Off

Bulk Read (Ep81h) max packet size: 1024
Bulk Write (Ep01h) max packet size: 1024

While the test is running:
Press 'Q' to quit
Press 'T' for test details
Press 'I' for status information
Press 'R' to reset averages

Press 'Q' to exit, any other key to begin..
Avg. Bytes/s: 292996045.43 Transfers: 4146 Bytes/s: 292996045.43
Avg. Bytes/s: 282608792.44 Transfers: 8314 Bytes/s: 272982151.14
Avg. Bytes/s: 283038163.13 Transfers: 12651 Bytes/s: 283864922.35
Avg. Bytes/s: 286039681.66 Transfers: 17160 Bytes/s: 294811405.17
Avg. Bytes/s: 286925700.90 Transfers: 21593 Bytes/s: 290407828.95
Avg. Bytes/s: 287219544.40 Transfers: 25999 Bytes/s: 288668364.04
Avg. Bytes/s: 288399570.98 Transfers: 30507 Bytes/s: 295398949.57
Avg. Bytes/s: 288607508.22 Transfers: 34948 Bytes/s: 290044056.90
Avg. Bytes/s: 288482105.26 Transfers: 39336 Bytes/s: 287487216.77
Avg. Bytes/s: 289061140.16 Transfers: 43828 Bytes/s: 294232769.02
Avg. Bytes/s: 290082763.74 Transfers: 48411 Bytes/s: 300230225.01
Avg. Bytes/s: 290219994.35 Transfers: 52867 Bytes/s: 291719310.22
Avg. Bytes/s: 291168280.68 Transfers: 57484 Bytes/s: 302485548.25
Avg. Bytes/s: 291603326.04 Transfers: 62021 Bytes/s: 297230126.38
qAvg. Bytes/s: 291859411.24 Transfers: 66538 Bytes/s: 295421649.76
stopped Ep81h thread.
stopped Ep01h thread.
Loop Test Information
	Vid / Pid       : 04B4h / 00F1h
	Interface #     : 00h
	Alt Interface # : 00h
	Priority        : Normal
	Buffer Size     : 65536
	Buffer Count    : 4
	Display Refresh : 1000 (ms)
	Transfer Timeout: 5000 (ms)
	Retry Count     : 0
	Verify Data     : Off

Bulk Read (Ep81h) max packet size: 1024
	Total Bytes     : 4360699904
	Total Transfers : 66539
	Avg. Bytes/sec  : 291857437.27
	Elapsed Time    : 14.94 seconds

Bulk Write (Ep01h) max packet size: 1024
	Total Bytes     : 2919825408
	Total Transfers : 44553
	Avg. Bytes/sec  : 195424164.60
	Elapsed Time    : 14.94 seconds

Press any key to exit..


```

**Comment by mcuee** [comment]
From here we know it is good to have some good host codes to test isoc transfer.
https://github.com/libusb/libusb/issues/1104

Tomord Volden has extend the iso-test codes from Wander Lairson Costa.
https://github.com/tormodvolden/isochronous

He has also extend the sam3u_benchmark example to stm32_isoc here (along with mod FW for STM32F1)
https://libusb.sourceforge.io/stm32_isoc/

**Comment by mcuee** [comment]
Unlikely this will happen any time soon. Close this one for now.
